File: C++/Path Sum II.cpp
#include<iostream>
#include<vector>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

void findpath(vector<vector<int> >&vec,vector<int> &st,TreeNode *root,int sum,int cur)
    {
        if(root->left==NULL&&root->right==NULL&&cur+root->val==sum)
            {
                st.push_back(root->val);
                vec.push_back(st);
                st.pop_back();
                return;
            }
        if(root->left==NULL&&root->right==NULL&&cur+root->val==sum)
            {
                vec.pop_back();
                return;
            }
        if(root->left!=NULL)
            {
                st.push_back(root->val);
                findpath(vec, st, root->left, sum, cur+root->val);
                st.pop_back();
            }
        if(root->right!=NULL)
            {
                st.push_back(root->val);
                findpath(vec, st, root->right, sum, cur+root->val);
                st.pop_back();
            }
        return ;
    }

vector<vector<int> > pathSum(TreeNode *root, int sum)
    {
        vector<vector<int> > vec;
        if(root==NULL) return vec;
        vector<int> st;
        findpath(vec, st, root, sum, 0);
        return vec;
    }

File: Python/Path Sum II.py
#!/usr/bin/env python

import copy

# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @param sum, an integer
    # @return a list of lists of integers
    def makepathSum(self, root, sum, seq, cnt):
        if root == None:
            return seq
        if root.left == None and root.right == None and root.val == sum:
            cnt.append(root.val)
            seq.append(copy.copy(cnt))
            cnt.pop()
            return seq
        if root.left != None:
            cnt.append(root.val)
            seq = self.makepathSum(root.left, sum - root.val, seq, cnt)
            cnt.pop()
        if root.right != None:
            cnt.append(root.val)
            seq = self.makepathSum(root.right, sum - root.val, seq, cnt)
            cnt.pop()
        return seq
    def pathSum(self, root, sum):
        seq = []
        cnt = []
        seq = self.makepathSum(root, sum, seq, cnt)
        return seq